src/smartpeak/source/ui/Widget.cpp

```diff
@@ -175,9 +176,20 @@ namespace SmartPeak
 
   void ParametersTableWidget::draw()
   {
-    if (headers_.size() <= 0)
+    if (application_handler_.sequenceHandler_.getSequence().size() == 0) {
       return;
-
+    }
+    std::vector<std::string> command_names = application_handler_.sequenceHandler_.getWorkflow();
+    ParameterSet user_parameters = application_handler_.sequenceHandler_.getSequence().at(0).getRawData().getParameters();
+    if ((command_names != input_command_names_) || (user_parameters != input_user_parameters_))
+    {
+      BuildCommandsFromNames buildCommandsFromNames(application_handler_);
+      buildCommandsFromNames.names_ = command_names;
+      buildCommandsFromNames.process(); 
+      session_handler_.getParametersTable(user_parameters, buildCommandsFromNames.commands_, headers_, body_);
```

Comment:
This change moves the logic for when to decide when to update the parameters table to the Widget.  Do you think this is the strategy that we will move forward with in the future for the rest of the widgets?  My concern is that the widget is not just displaying the data but also deciding when to update the data.  Do you see any issues in regards to not separating the logic of updating the data and displaying the data?

Reply:
Yes, good point. The thing is when I implemented this, there was no way to know that the data, here, the couple (parameter, workflow) has changed. While I used workaround for the chromatogram (wait that the workflow finished to run, from GUI.cpp), here it's more complicated: 

1. Parameter can be load, maybe later edited, 
2. and more problematic there the workflow can be load, and edited. And we access directly to the vector to push items to this.

To make it proper, we need to introduce observers to notify the view of the change from the model. The model needs means to know the data has changed (that is: not accessing vectors, or, give a mean to notify changes).

So I can proceed and introduce these changes with the opportunity of this PR. on this context, I see 2 options of doing that:

1. encapsulate the maps and vectors that we access directly (workflow, but in this case also Parameter fields) so that when the change the view is notified. It could be a little bit too much.
2.  introduce notification methods in the model (sessionhandler). we still access to the vectors, maps, parameter fields, but the call will have to call notification function from the model, when appropriate. It would reduce the changes, and may be let open the option to do 1) later.
 so for example, for 2) we will do 
```
worflow_.push("MyCommand")
sessionHandler.notifyWorkflowChange() // sessionhandler will notify observer (the views)
```

Reply:
Please proceed with the solution that you think is best.  I agree that we need an overhaul of how and when we decide different views should be updated.  I think the `parameters` provides a good use case for beginning to develop such a solution/strategy.  If you would like to discuss the options or next steps further please just let me know.  Other than that I will wait until the next review request to continue our discussion after you have had time to test out and implement the options that you see as optimal.